File: wagerproof-android-native/core/stores/src/main/kotlin/com/wagerproof/core/stores/FollowedAgentsStore.kt
package com.wagerproof.core.stores

import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.wagerproof.core.models.Agent
import com.wagerproof.core.models.AgentPerformance
import com.wagerproof.core.models.AgentSport
import com.wagerproof.core.models.AgentWithPerformance
import com.wagerproof.core.services.SupabaseClients
import com.wagerproof.core.services.runCatchingCancellable
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Port of iOS `FollowedAgentsStore.swift` (doc §8.10). Read-path for the public
 * agents the user follows (`user_avatar_follows` join on the MAIN Supabase
 * project). Mutations live elsewhere (AgentChatService.setFollow).
 *
 * Backs the Following rail + See All sheet at the top of My Agents, so the
 * projection has to carry everything those surfaces render: the profile columns
 * for the tile, `last_generated_at` for the unread dot, and a second batched
 * `avatar_performance_cache` read for the ±5 streak badge and the sheet's
 * record/units row. A narrower select is what left the rail unbuildable before.
 */
@Stable
class FollowedAgentsStore {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    /** Coalesces concurrent refreshes (rail appear + follow toggle firing together). */
    private val inFlight = InFlightTaskRegistry<RefreshKey>(scope)

    /** Distinguishes sign-out/re-auth even when the same user id returns. */
    private var bindingGeneration = 0L

    private data class RefreshKey(val userId: String, val bindingGeneration: Long)

    /** One row of follow data. Mirrors the RN `FollowedAgent` interface. */
    data class FollowedAgent(
        val avatarId: String,
        val name: String,
        val avatarEmoji: String,
        val avatarColor: String,
        val isFavorite: Boolean,
        val lastGeneratedAt: String? = null,
        val userId: String = "",
        val preferredSports: List<AgentSport> = emptyList(),
        val performance: AgentPerformance? = null,
    ) {
        /**
         * Adapter so the See All sheet can reuse `AgentRowCard` unchanged. The
         * synthesized Agent is a display shell: only the fields the row renders
         * are real, and `isPublic` is true because you can only follow a public
         * agent.
         */
        val agentWithPerformance: AgentWithPerformance
            get() = AgentWithPerformance(
                agent = Agent(
                    id = avatarId,
                    userId = userId,
                    name = name,
                    avatarEmoji = avatarEmoji,
                    avatarColor = avatarColor,
                    preferredSports = preferredSports,
                    isPublic = true,
                    createdAt = "",
                    lastGeneratedAt = lastGeneratedAt,
                ),
                performance = performance,
            )
    }

    var follows by mutableStateOf<List<FollowedAgent>>(emptyList()); private set
    var loadState by mutableStateOf<LoadState>(LoadState.Idle); private set
    var userId by mutableStateOf<String?>(null); private set

    fun bind(userId: String?) {
        if (userId == this.userId) return
        bindingGeneration += 1
        this.userId = userId
        // Requests run on the store scope and deliberately outlive their screen
        // caller. Once identity changes that behavior is no longer safe: cancel
        // the old work, then reject any non-cooperative completion via the key
        // checks in [load].
        inFlight.cancelAll()
        follows = emptyList()
        loadState = LoadState.Idle
    }

    /**
     * Fetch the follow set: `user_avatar_follows` joined to `avatar_profiles`,
     * then one batched `avatar_performance_cache` read for the same ids.
     *
     * Two queries, not a three-table join: `avatar_performance_cache` is a
     * separate cache table with no FK PostgREST can traverse from the join row.
     */
    suspend fun refresh() {
        val uid = userId ?: return
        val key = RefreshKey(uid, bindingGeneration)
        inFlight.run(key) { load(key) }
    }

    private suspend fun load(key: RefreshKey) {
        if (!isCurrent(key)) return
        loadState = LoadState.Loading
        runCatchingCancellable {
            val rows = SupabaseClients.main.from("user_avatar_follows")
                .select(
                    columns = Columns.raw(
                        "avatar_id, is_favorite, avatar_profiles(user_id, name, avatar_emoji, " +
                            "avatar_color, preferred_sports, last_generated_at)",
                    ),
                ) {
                    filter { eq("user_id", key.userId) }
                }
                .decodeList<FollowRow>()
            val ids = rows.map { it.avatarId }
            // Skip the second round-trip entirely when nothing is followed —
            // an `in` filter on an empty list is a wasted request. A perf-cache
            // failure never blanks the rail: rows just render performance-less
            // (same defensive shape as AgentService.fetchUserAgents).
            val performances = if (ids.isEmpty()) {
                emptyList()
            } else {
                runCatchingCancellable {
                    SupabaseClients.main.from("avatar_performance_cache")
                        .select(columns = Columns.raw(PERFORMANCE_COLUMNS)) {
                            filter { isIn("avatar_id", ids) }
                        }
                        .decodeList<AgentPerformance>()
                }.getOrDefault(emptyList())
            }
            rows to performances.associateBy { it.avatarId }
        }.onSuccess { (rows, performanceById) ->
            if (!isCurrent(key)) return@onSuccess
            follows = rows.map { row ->
                FollowedAgent(
                    avatarId = row.avatarId,
                    name = row.avatarProfiles?.name ?: "Unknown",
                    avatarEmoji = row.avatarProfiles?.avatarEmoji ?: "🤖",
                    avatarColor = row.avatarProfiles?.avatarColor ?: "#6366f1",
                    isFavorite = row.isFavorite ?: false,
                    lastGeneratedAt = row.avatarProfiles?.lastGeneratedAt,
                    userId = row.avatarProfiles?.userId ?: "",
                    preferredSports = row.avatarProfiles?.preferredSports ?: emptyList(),
                    performance = performanceById[row.avatarId],
                )
            }
            loadState = LoadState.Loaded
        }.onFailure {
            if (!isCurrent(key)) return@onFailure
            loadState = LoadState.Failed(it.message.orEmpty().ifEmpty { "Unknown error" })
        }
    }

    private fun isCurrent(key: RefreshKey): Boolean =
        userId == key.userId && bindingGeneration == key.bindingGeneration

    /**
     * Drop a row locally so an unfollow feels instant. The authoritative list
     * still comes from the next [refresh]; this only avoids a visible lag
     * between the swipe and the round-trip.
     */
    fun removeLocally(agentId: String) {
        follows = follows.filterNot { it.avatarId == agentId }
    }

    fun close() = scope.cancel()

    private companion object {
        /**
         * Explicit projection, mirroring AgentService: the JSONB `stats_by_*`
         * columns arrive as SQL NULL for agents with no graded picks, and a null
         * for a non-nullable Map field is a decode failure (a defaulted field
         * only covers an ABSENT key, not a present null).
         */
        const val PERFORMANCE_COLUMNS =
            "avatar_id, wins, losses, pushes, total_picks, win_rate, net_units, " +
                "current_streak, best_streak, worst_streak, last_calculated_at"
    }

    /** Internal join-row DTO. */
    @Serializable
    private data class FollowRow(
        @SerialName("avatar_id") val avatarId: String,
        @SerialName("is_favorite") val isFavorite: Boolean? = null,
        @SerialName("avatar_profiles") val avatarProfiles: NestedProfile? = null,
    ) {
        @Serializable
        data class NestedProfile(
            val name: String? = null,
            @SerialName("user_id") val userId: String? = null,
            @SerialName("avatar_emoji") val avatarEmoji: String? = null,
            @SerialName("avatar_color") val avatarColor: String? = null,
            @SerialName("preferred_sports") val preferredSports: List<AgentSport>? = null,
            @SerialName("last_generated_at") val lastGeneratedAt: String? = null,
        )
    }
}

File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofStores/FollowedAgentsStore.swift
import Foundation
import Observation
import Supabase
import WagerproofModels
import WagerproofServices

/// `FollowedAgentsStore` ports `wagerproof-mobile/hooks/useFollowedAgents.ts`.
/// Tracks the public agents the current user follows by reading the
/// `user_avatar_follows` join table on the main Supabase project.
///
/// Follow/unfollow mutations themselves land in B16 with the Public Agent
/// Detail screen — this store only exposes the read path so the Top Picks
/// inner tab can filter to "Following" later. For B13 we expose the list.
@Observable
@MainActor
public final class FollowedAgentsStore {
    public enum LoadState: Equatable, Sendable {
        case idle
        case loading
        case loaded
        case failed(String)
    }

    /// One row of follow data. Mirrors the RN `FollowedAgent` interface.
    public struct FollowedAgent: Identifiable, Sendable, Hashable {
        public let avatarId: String
        public let name: String
        public let avatarEmoji: String
        public let avatarColor: String
        public let isFavorite: Bool
        public let lastGeneratedAt: String?
        public let userId: String
        public let preferredSports: [AgentSport]
        public let performance: AgentPerformance?

        public var id: String { avatarId }

        public init(
            avatarId: String,
            name: String,
            avatarEmoji: String,
            avatarColor: String,
            isFavorite: Bool,
            lastGeneratedAt: String? = nil,
            userId: String = "",
            preferredSports: [AgentSport] = [],
            performance: AgentPerformance? = nil
        ) {
            self.avatarId = avatarId
            self.name = name
            self.avatarEmoji = avatarEmoji
            self.avatarColor = avatarColor
            self.isFavorite = isFavorite
            self.lastGeneratedAt = lastGeneratedAt
            self.userId = userId
            self.preferredSports = preferredSports
            self.performance = performance
        }

        public var agentWithPerformance: AgentWithPerformance {
            AgentWithPerformance(
                agent: Agent(
                    id: avatarId,
                    userId: userId,
                    name: name,
                    avatarEmoji: avatarEmoji,
                    avatarColor: avatarColor,
                    preferredSports: preferredSports,
                    archetype: nil,
                    isPublic: true,
                    createdAt: "",
                    updatedAt: "",
                    lastGeneratedAt: lastGeneratedAt
                ),
                performance: performance
            )
        }
    }

    public private(set) var follows: [FollowedAgent] = []
    public private(set) var loadState: LoadState = .idle
    public private(set) var userId: String?

    public init() {}

    public func bind(userId: String?) {
        if userId == self.userId { return }
        self.userId = userId
        self.follows = []
        self.loadState = .idle
    }

    /// Fetch the follow set. Joins `user_avatar_follows` against
    /// `avatar_profiles(name, avatar_emoji, avatar_color)` — matches the RN
    /// `useFollowedAgents` query in `useFollowedAgents.ts:28-46`.
    public func refresh() async {
        guard let userId else { return }
        loadState = .loading
        do {
            let main = await MainSupabase.shared.client
            let rows: [FollowRow] = try await main
                .from("user_avatar_follows")
                .select("avatar_id, is_favorite, avatar_profiles(user_id, name, avatar_emoji, avatar_color, preferred_sports, last_generated_at)")
                .eq("user_id", value: userId)
                .execute()
                .value
            let ids = rows.map(\.avatarId)
            let performances: [AgentPerformance]
            if ids.isEmpty {
                performances = []
            } else {
                performances = try await main
                    .from("avatar_performance_cache")
                    .select()
                    .in("avatar_id", values: ids)
                    .execute()
                    .value
            }
            let performanceById = Dictionary(
                uniqueKeysWithValues: performances.map { ($0.avatarId, $0) }
            )
            self.follows = rows.map { row in
                FollowedAgent(
                    avatarId: row.avatarId,
                    name: row.avatarProfiles?.name ?? "Unknown",
                    avatarEmoji: row.avatarProfiles?.avatarEmoji ?? "\u{1F916}",
                    avatarColor: row.avatarProfiles?.avatarColor ?? "#6366f1",
                    isFavorite: row.isFavorite ?? false,
                    lastGeneratedAt: row.avatarProfiles?.lastGeneratedAt,
                    userId: row.avatarProfiles?.userId ?? "",
                    preferredSports: row.avatarProfiles?.preferredSports ?? [],
                    performance: performanceById[row.avatarId]
                )
            }
            self.loadState = .loaded
        } catch {
            self.loadState = .failed((error as NSError).localizedDescription)
        }
    }

    /// Internal join-row Codable.
    private struct FollowRow: Decodable {
        let avatarId: String
        let isFavorite: Bool?
        let avatarProfiles: NestedProfile?

        enum CodingKeys: String, CodingKey {
            case avatarId = "avatar_id"
            case isFavorite = "is_favorite"
            case avatarProfiles = "avatar_profiles"
        }

        struct NestedProfile: Decodable {
            let name: String?
            let userId: String?
            let avatarEmoji: String?
            let avatarColor: String?
            let preferredSports: [AgentSport]?
            let lastGeneratedAt: String?
            enum CodingKeys: String, CodingKey {
                case name
                case userId = "user_id"
                case avatarEmoji = "avatar_emoji"
                case avatarColor = "avatar_color"
                case preferredSports = "preferred_sports"
                case lastGeneratedAt = "last_generated_at"
            }
        }
    }
}
